qualcommax: ipq50xx: Update GE PHY driver and fix GE PHY and uniphy resets

georgemoussalem opened the pull request:
Two out of 6 related commits have been submitted to the new openwrt qca-ssdk repo.
Below is the right sequence.

**Commit 1**:
qualcommax: ipq50xx: update patches for IPQ5018 GE PHY
    
Use latest patches sent upstream for review for IPQ5018 GE PHY support:
- Move enablement of the LDO controller to the mdio-ipq4019 driver away
  from the CMN PLL driver
- Remove the different patches to add CDT, MSE, AZ, and DAC support they
  are all contained in the upstreamed driver.
    
Accordingly, also set the right property in the DTS for Linksys SPNMX56
to set the right DAC values to accommodate for the short cable length.

**Commit 2**:
Commit 1 in https://github.com/openwrt/qca-ssdk/pull/1

**Commit 3**:
qualcommax: ipq50xx: remove named clock references for GE PHY
    
With completely carving out GE PHY out of the QCA-SSDK, the named clock
references to the GE PHY RX and TX clocks are no longer needed.
So, let's revert to using the DT indices as per the upstream GCC driver.

**Commit 4**:
qualcommax: ipq50xx: prepare for UNIPHY reset fix

Ahead of the actual fix in both the GCC and QCA-SSDK, add the required
AHB reset so it can be picked up by updated QCA-SSDK. This is needed
as the SSDK needs to use different resets depending on the link
architecture. If it's a fixed link, AHB needs to be reset. In a phy to
phy link setup (such as QCA8081), SYS, RX, and TX need to be reset using
one reset with a bitmask in the GCC (GCC_UNIPHY_SOFT_RESET).

**Commit 5**:
Commit 2 in https://github.com/openwrt/qca-ssdk/pull/1

**Commit 6**:
qualcommax: ipq50xx: fix uniphy soft reset issue
    
The resets in the GCC of the uniphy found in the IPQ5018 SoC are
incorrect which broke the ability to shift between 1G and 2.5G link
speeds. So let's correct the resets based on below two downstream
commits.
    
In a seperate and prequisite PR to the QCA-SSDK repo, logic has been
implemented to select the right reset based on the link setup so fixed
link scenarios don't break.
    
Link: https://git.codelinaro.org/clo/qsdk/oss/kernel/linux-ipq-5.4/-/commit/00743c3e82fa87cba4460e7a2ba32f473a9ce932
Link: https://git.codelinaro.org/clo/qsdk/oss/kernel/linux-ipq-5.4/-/commit/036bdc62aca561e8ff94a29c447fc400de2b7304

georgemoussalem committed:
qualcommax: ipq50xx: prepare for UNIPHY reset fix

Ahead of the actual fix in both the GCC and QCA-SSDK, add the required
AHB reset so it can be picked up by updated QCA-SSDK. This is needed
as the SSDK needs to use different resets depending on the link
architecture. If it's a fixed link, AHB needs to be reset. In a phy to
phy link setup (such as QCA8081), SYS, RX, and TX need to be reset using
one reset with a bitmask in the GCC (GCC_UNIPHY_SOFT_RESET).

Signed-off-by: George Moussalem <george.moussalem@outlook.com>
Link: https://github.com/openwrt/openwrt/pull/18774
Signed-off-by: Robert Marko <robimarko@gmail.com>

hzyitc commented:
> Commit 1:
> qualcommax: ipq50xx: update IPQ5018 Internal PHY driver
>
> Update the GE PHY driver and move enablement of the LDO controller to
> this driver away from the CMN PLL driver as it is a component of this
> phy. Let's also remove the different patches to add CDT, MSE, AZ, and
> DAC support as this is all in this driver now.

IIRC, LDO is for CMN, not only PHY. All network devices require it.

> Commit 2:
> ...
> Commit 5:
> The resets in the GE PHY node has been updated as the fixed reset uses
> a bitmask to cover the ones replaced simulateously.

I don't think it's a good idea to merge lots of resets into one. For now, it's easy to understand. What we need to do is leting SSDK reset all of these. `reset_control_bulk_get_*` serial functions are what you need.

georgemoussalem commented:
> > Commit 1:
> > qualcommax: ipq50xx: update IPQ5018 Internal PHY driver
> > Update the GE PHY driver and move enablement of the LDO controller to
> > this driver away from the CMN PLL driver as it is a component of this
> > phy. Let's also remove the different patches to add CDT, MSE, AZ, and
> > DAC support as this is all in this driver now.
> 
> IIRC, LDO is for CMN, not only PHY. All network devices require it.

no, I thought so too, but it is a component of the internal ge PHY only (confirmed by qcom)

> 
> > Commit 2:
> > ...
> > Commit 5:
> > The resets in the GE PHY node has been updated as the fixed reset uses
> > a bitmask to cover the ones replaced simulateously.
> 
> I don't think it's a good idea to merge lots of resets into one. For now, it's easy to understand. What we need to do is leting SSDK reset all of these. `reset_control_bulk_get_*` serial functions are what you need.

I don’t really have a preference on listing all the resets or the misc reset in the dts, but that’s how the resets are defined in the downstream gcc.
But the PHY driver itself should do the resets, it should be self sufficient before sending I’ll send it upstream. The SSDK causes weird behavior and duplicates the resets as you can see in the associated PR to the qca-ssdk repo. This PR guards off the manual resets in the SSDK for the PHY driver to be in control.

robimarko commented:
Would you mind if we would hold-off this until 6.12 lands?

I plan to merge it quickly, we can always fixup stuff while its testing.

georgemoussalem commented:
Sure, sounds good.

georgemoussalem committed:
qualcommax: ipq50xx: update patches for IPQ5018 GE PHY

Use latest patches sent upstream for review for IPQ5018 GE PHY support:
- Move enablement of the LDO controller to the mdio-ipq4019 driver away
  from the CMN PLL driver
- Remove the different patches to add CDT, MSE, AZ, and DAC support they
  are all contained in the upstreamed driver.

Accordingly, also set the right property in the DTS for Linksys SPNMX56
to set the right DAC values to accommodate for the short cable length.

Signed-off-by: George Moussalem <george.moussalem@outlook.com>
Link: https://github.com/openwrt/openwrt/pull/18774
Signed-off-by: Robert Marko <robimarko@gmail.com>

georgemoussalem committed:
qualcommax: ipq50xx: fix uniphy soft reset issue

The resets in the GCC of the uniphy found in the IPQ5018 SoC are
incorrect which broke the ability to shift between 1G and 2.5G link
speeds. So let's correct the resets based on below two downstream
commits.

In a seperate and prequisite PR to the QCA-SSDK repo, logic has been
implemented to select the right reset based on the link setup so fixed
link scenarios don't break.

Signed-off-by: George Moussalem <george.moussalem@outlook.com>
Link: https://github.com/openwrt/openwrt/pull/18774
Signed-off-by: Robert Marko <robimarko@gmail.com>

georgemoussalem committed:
qualcommax: ipq50xx: remove named clock references for GE PHY

With completely carving out GE PHY out of the QCA-SSDK, the named clock
references to the GE PHY RX and TX clocks are no longer needed.
So, let's revert to using the DT indices as per the upstream GCC driver.

Signed-off-by: George Moussalem <george.moussalem@outlook.com>
Link: https://github.com/openwrt/openwrt/pull/18774
Signed-off-by: Robert Marko <robimarko@gmail.com>

georgemoussalem committed:
qca-ssdk: update to Git HEAD (2025-05-30)

3423fdacac63 ssdk_clk: Remove MP_PHY clocks resets and init logic

Signed-off-by: George Moussalem <george.moussalem@outlook.com>
Link: https://github.com/openwrt/openwrt/pull/18774
Signed-off-by: Robert Marko <robimarko@gmail.com>

georgemoussalem committed:
qca-ssdk: update to Git HEAD (2025-05-30)

446db12b1fd3 MP: fix uniphy reset in phy to phy link scenario

Signed-off-by: George Moussalem <george.moussalem@outlook.com>
Link: https://github.com/openwrt/openwrt/pull/18774
Signed-off-by: Robert Marko <robimarko@gmail.com>

robimarko commented:
Thanks! Rebased on top of main and merged!
